lib/chat/db/scope/invitations_handshaker.ex

```diff
@@ -0,0 +1,265 @@
+defmodule Chat.Db.Scope.InvitationsHandshaker do
+  @moduledoc """
+  Process cargo room invitations by handshaker
+  """
+
+  import Chat.Db.Scope.Utils
+  alias Chat.Dialogs.Dialog
+  alias Chat.Dialogs.Message
+
+  def build_initial_context(snap) do
+    %{}
+    |> put_room_invitations_content(snap, nil)
+    |> put_invitations_keymap()
+    |> put_invitations_messages(snap)
+    |> clear_context_preparation()
+  end
+
+  def put_room_invitations_content(context, snap, keys) do
+    [room_invite_index, _room_invite_keys, room_invites] =
+      snap
+      |> fetch_index_and_records(
+        keys,
+        "room_invite",
+        min_key: {:room_invite_index, 0, 0},
+        max_key: {:"room_invite_index\0", 0, 0},
+        reader_hash_getter: fn {:room_invite_index, reader_key, _invite_key} -> reader_key end,
+        record_key_getter: fn {:room_invite_index, _reader_key, invite_key} -> invite_key end
+      )
+
+    Map.merge(context, %{room_invite_indexes: room_invite_index, room_invites: room_invites})
+  end
+
+  def fetch_index_and_records(snap, pub_keys, record_name, opts) do
+    reader_hash_getter = opts[:reader_hash_getter]
+    record_key_getter = opts[:record_key_getter]
+
+    index =
+      snap
+      |> db_keys_stream(opts[:min_key], opts[:max_key])
+      |> filter_invitations_indexes(reader_hash_getter, pub_keys)
+      |> MapSet.new()
+
+    keys =
+      index
+      |> Enum.map(&record_key_getter.(&1))
+      |> MapSet.new()
+
+    records =
+      snap
+      |> db_keys_stream({:"#{record_name}", 0}, {:"#{record_name}\0", 0})
+      |> filter_invitations_records(keys)
+      |> MapSet.new()
+
+    [index, keys, records]
+  end
+
+  def handshake_dialogs_cycle(context, pub_keys) do
+    move = 0
+
+    context
+    |> start_invitations_queue(pub_keys)
+    |> execute_handshake_cycle(move)
+  end
+
+  def execute_handshake_cycle(
+        %{invitations_messages: messages, invitations_queue: queue} = context,
+        move
+      ) do
+    context =
+      if not Enum.empty?(messages) && move <= 2 do
+        queue
+        |> List.first()
+        |> get_messages_sender_keys(move)
+        |> fetch_cycle_invites(context)
+        |> put_invitations_queue(context)
+        |> remove_queued_invitations_messages()
+      else
+        context
+      end
+
+    move = move + 1
+    if move == 3, do: context, else: execute_handshake_cycle(context, move)
+  end
```

Comment:
What is the logic behind `move`?

Reply:
Move is describe a current cycling number of getting handshake invites. Based on it i get messages direction and catch when should cycle be stopped. It starts from 0 and get invitations messages to checkpoints, then use its senders for each next iteration

Reply:
Ok i find a way to not use this magic variable

Reply:
Refactored